subsys/nfc/t4t/isodep.c

```diff
@@ -0,0 +1,877 @@
+			if (len > index) {
+				LOG_DBG("S-Frame S(DESELECT), invalid length received.");
+				return -NFC_T4T_ISODEP_SYNTAX_ERROR;
+			}
+
+			atomic_set(&t4t_isodep.state, ISODEP_STATE_INITIALIZED);
+
+			return 0;
+		} else {
+			return -NFC_T4T_ISODEP_SYNTAX_ERROR;
+		}
+	}
+
+	/* Frame Waiting Time Extension Request contains 1 - byte
+	 * INF field. Check if length of Request is correct.
+	 */
+	if (len > (index + 1)) {
+		LOG_DBG("S-Frame S(WTX), invalid length received.");
+		return -EINVAL;
+	}
+
+	wtxm = data[index];
+	wtxm &= ISODEP_WTXM_MASK;
+
+	/* Check if WTXM is in valid range. */
+	if ((wtxm < ISODEP_WTXM_MIN) || (wtxm > ISODEP_WTXM_MAX)) {
+		return -NFC_T4T_ISODEP_SYNTAX_ERROR;
+	}
+
+	/* Prepare S(WTX) response. */
+	index = 0;
+	tx_data = t4t_isodep.tx_data.data;
+
+	tx_data[index] = ISODEP_S_BLOCK | S_BLOCK_WTX_MASK;
+
+	if (did_include(&tx_data[index])) {
+		index++;
+	}
+
+	index++;
+
+	/* Respond with the same WTXM. */
+	tx_data[index] = wtxm;
+	index++;
+
+	t4t_isodep.tx_data.len = index;
+
+	/* Calculate new Frame Delay Time value. */
+	fdt  = t4t_isodep.tag.fwt * wtxm + T4T_FWT_DELTA + NFCA_T4T_FWT_T_FC;
+
+	/* In case of error, remember last send frame type. */
+	t4t_isodep.err_status.last_frame = ISODEP_FRAME_WTX_RESPONSE;
+
+	if (t4t_isodep_cb->ready_to_send) {
+		t4t_isodep_cb->ready_to_send(tx_data, t4t_isodep.tx_data.len, fdt);
+	}
+
+	return 0;
+}
+
+static int isodep_r_frame_handle(const u8_t *data, size_t len)
+{
+	size_t index = 0;
+	u8_t r_block;
+	u32_t fdt;
+
+	r_block = data[index];
+	index++;
+
+	if (r_block & I_BLOCK_DID_BIT) {
+		index++;
+	}
+
+	if (len > index) {
+		return -NFC_T4T_ISODEP_SYNTAX_ERROR;
+	}
+
+	/* Check if ACK or NAK */
+	if (r_block & R_BLOCK_NAK)  {
+		LOG_DBG("R-frame NAK received.");
+		return -NFC_T4T_ISODEP_SEMATIC_ERROR;
+	}
+
+	LOG_DBG("R-frame ACK received.");
+
+	/* Check if Reader/Writer is in chaining state. */
+	if (!t4t_isodep.chaining) {
+		return -NFC_T4T_ISODEP_SEMATIC_ERROR;
+	}
+
+	/* ACK Received. Check if current block number is equal to receieved. */
+	if ((r_block & ISODEP_BLOCK_NUM_MASK) == t4t_isodep.block_num) {
+		/* Toggle block number */
+		block_num_toggle();
+
+		isodep_chunk_send();
+	} else {
+		/* Retransmit last data block */
+		if (t4t_isodep.retransmit_cnt > ISODEP_MAX_RETRANSMIT_CNT) {
+			/* If to much retransmission, return error. */
+			return -NFC_T4T_ISODEP_SEMATIC_ERROR;
+		}
+
+		fdt = t4t_isodep.tag.fwt + T4T_FWT_DELTA + NFCA_T4T_FWT_T_FC;
+
+		if (t4t_isodep_cb->ready_to_send) {
+			t4t_isodep_cb->ready_to_send(t4t_isodep.tx_data.data,
+						     t4t_isodep.tx_data.len, fdt);
+
+			t4t_isodep.retransmit_cnt++;
+		}
+	}
+
+	return 0;
+}
+
+static int isodep_i_frame_handle(const u8_t *data, size_t len)
+{
+	size_t index = 0;
+	u8_t i_block;
+
+	i_block = data[index];
+	index++;
+
+	if (i_block & I_BLOCK_DID_BIT) {
+		index++;
+	}
+
+	if (i_block & I_BLOCK_NAD_BIT) {
+		index++;
+	}
+
+	LOG_DBG("Valid I-Frame received");
```

Comment:
what if len < 3?

Reply:
Len can be < 3. Did and Nad field in most cases are not included.

Reply:
I added additional check if `index < len`